saas-starter-ts/backend/subscription/stripe.ts

```diff
@@ -0,0 +1,258 @@
+        VALUES (
+          ${subscription.id as string},
+          ${subscription.customer as string},
+          ${subscription.items.data[0].plan.id as string},
+          ${subscription.status as string},
+          ${new Date(subscription.created * 1000)},
+          ${new Date(subscription.created * 1000)}
+        )
+        ON CONFLICT (stripe_id) DO UPDATE SET
+          stripe_customer_id = ${subscription.customer as string},
+          price_id = ${subscription.items.data[0].plan.id as string},
+          status = ${subscription.status as string},
+          updated_at = ${new Date(subscription.created * 1000)}
+        `;
+          break;
+        }
+      }
+
+      res.write(JSON.stringify({ received: true }));
+      res.writeHead(200);
+      res.end();
+    } catch (error) {
+      log.error("stripe webhook error", { error });
+
+      res.write(JSON.stringify({ received: false }));
+      res.writeHead(500);
+      res.end();
+    }
+  },
+);
+
+interface CreateCheckoutSessionParams {
+  priceId: string;
+}
+
+interface CreateCheckoutSessionResponse {
+  // The URL to redirect to
+  url: string;
+}
+
+// Generate a checkout session
+export const createCheckoutSession = api(
+  {
+    method: "POST",
+    path: "/stripe/checkout-session",
+    expose: true,
+    auth: true,
+  },
+  async (
+    params: CreateCheckoutSessionParams,
+  ): Promise<CreateCheckoutSessionResponse> => {
+    const authData = getAuthData();
+    if (!authData) {
+      throw APIError.unauthenticated("user not authenticated");
+    }
+
+    // Try to get the organization's stripe customer ID
+    // This is stored as organization metadata, please see https://clerk.com/docs/organizations/metadata for more information.
+    const org = await clerk.organizations.getOrganization({
+      organizationId: authData.orgID,
+    });
+    let stripeCustomerId = org.privateMetadata.stripeCustomerId;
+
+    // If the organization doesn't have a Stripe customer ID, create one
+    if (!stripeCustomerId) {
+      const user = await clerk.users.getUser(authData.userID);
+
+      const customer = await stripe.customers.create({
+        name: org.name,
+        email: user.primaryEmailAddress?.emailAddress,
+        metadata: {
+          orgId: authData.orgID,
+        }
+      });
+      await clerk.organizations.updateOrganizationMetadata(authData.orgID, {
+        privateMetadata: {
+          stripeCustomerId: customer.id,
+        },
+      });
+      stripeCustomerId = customer.id;
+    }
+
+    const session = await stripe.checkout.sessions.create({
+      billing_address_collection: "auto",
+      customer: stripeCustomerId,
+      client_reference_id: authData.orgID,
+      line_items: [
+        {
+          price: params.priceId,
+          // For metered billing, do not pass quantity
+          quantity: 1,
+        },
+      ],
+      mode: "subscription",
+      success_url:
```

Comment:
These hard-coded URLs should be moved to a configuration file and documented how to tweak them